src/ixdat/calculators/packs.py

```diff
@@ -0,0 +1,559 @@
+            payload = item["payload"]
+
+            # Integrity check (best-effort)
+            h_expected = item.get("hash")
+            h_actual = _sha256_of_jsonable(payload)
+            if h_expected and h_expected != h_actual:
+                raise ValueError(
+                    f"Integrity check failed for {class_path}: hash mismatch"
+                )
+
+            CalCls = _import_from_class_path(class_path)
+
+            # Always use CalculatorPack’s deserializer
+            cal = self._deserialize_calculator(CalCls, payload)
+            calculators.append(cal)
+
+        self._calculators = calculators
+
+        return list(self._calculators)
+
+    def validate(self, *, strict: bool = True) -> Tuple[bool, List[str]]:
+        """
+        Validate the internal structure and per-calculator integrity hashes.
+
+        Parameters:
+            strict: If True, treat any discrepancy as an error;
+            otherwise aggregate warnings.
+
+        Returns:
+            (ok, messages): ok=True if everything looks good
+        """
+        msgs: List[str] = []
+        ok = True
+
+        if self.schema_version != self.SCHEMA_VERSION:
+            ok = False
+            msgs.append(
+                f"Schema version mismatch: pack={self.schema_version}, "
+                f"supported={self.SCHEMA_VERSION}"
+            )
+
+        if not self._items:
+            msgs.append("Pack contains no calculator entries")
+            return (ok and not strict, msgs)
+
+        for item in self._items:
+            cp = item.get("class_path")
+            payload = item.get("payload")
+            if not cp or not isinstance(cp, str):
+                ok = False
+                msgs.append("Entry missing valid 'class_path'")
+                if strict:
+                    continue
+            if not isinstance(payload, dict):
+                ok = False
+                msgs.append(f"{cp}: payload is not a dict")
+                if strict:
+                    continue
+            h_expected = item.get("hash")
+            if h_expected:
+                h_actual = _sha256_of_jsonable(payload)
+                if h_expected != h_actual:
+                    ok = False
+                    msgs.append(f"{cp}: integrity hash mismatch")
+
+        if strict and not ok:
+            return (False, msgs)
+        return (ok, msgs)
+
+    def attach_to(
+        self,
+        measurement: Any,
+        *,
+        on_conflict: str = "replace",
+    ) -> List[Any]:
+        """
+        Attach all calculators in the pack to a Measurement.
+
+        Parameters:
+            measurement: Target Measurement (must provide `.add_calculator(calculator)`).
+            on_conflict:
+                "replace" (default): replace existing calculator of same class+name
+                "skip": keep existing one, skip new
+                "duplicate": add new one and allow duplicates (caller handles names)
```

Comment:
"duplicate" effectively will require that the calculators can be added. `Measurement.calculators` uses [https://github.com/ixdat/ixdat/blob/40b7ef213d2e013473a500c709d2714de8c6a915/src/ixdat/measurement_base.py#L525](Measurement.consolidate_calculators), which combines calculators of the same type with their `__add__` method. This can make complete sense when e.g. you have an ECCalibration with the electrode area and another ECCalibraiton with the RE potential, or MSCalibrations with sensitivity factors for different molecules. But yeah I guess then they shouldn't have the same names...
So this isn't a request for a change, just something I wanted you to know.